### PR TITLE
refactor: Rework to be proper abstraction of a SecretStore and add SecretStoreClient from edgex-go

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #87

## What is the new behavior?
No longer leak that Vault is the only implementation
Added `Type` to the configuration and factory method, so it will error if not set
Move SecretStoreClient from edgex-go in to this abstraction and added the `EnableConsulSecretEngine` interface needed for `Secure Consul`
Now have to interfaces that the vault wrapper implements, which are SecretsClient & SecretStoreClient

SecretStoreClient code originated from here: https://github.com/edgexfoundry/edgex-go/tree/master/internal/security/secretstoreclient.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

BREAKING CHANGE: All existing SecretStore configuration must add `Type = 'vault'`

## Are there any new imports or modules? If so, what are they used for and why?

no

## Other information

### CODE REVIEW

File: internal/pkg/vault/management.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		},
+	}
+
+	_, err := c.doRequest(RequestArgs{
+		AuthToken:            token,
+		Method:               http.MethodPost,
+		Path:                 urlPath,
+		JSONObject:           parameters,
+		BodyReader:           nil,
+		OperationDescription: "update mounts",
```
Comment (jim-wang-intel) [inline internal/pkg/vault/management.go:142]:
```suggestion
		OperationDescription: "update mounts for KV",
```

File: internal/pkg/vault/management.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+func (c *Client) CheckSecretEngineInstalled(token string, mountPoint string, engine string) (bool, error) {
+	var response ListSecretEnginesResponse
+
+	_, err := c.doRequest(RequestArgs{
+		AuthToken:            token,
+		Method:               http.MethodGet,
+		Path:                 MountsAPI,
+		JSONObject:           nil,
+		BodyReader:           nil,
+		OperationDescription: "query mounts for Consul",
```
Comment (jim-wang-intel) [inline internal/pkg/vault/management.go:183]: Please update the description based on the engine type:
```suggestion
		OperationDescription: "query mounts for "+engine,
```

File: internal/pkg/vault/managment_test.go [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+	defer ts.Close()
+
+	client := createClient(t, ts.URL, mockLogger)
+
+	// Act
+	installed, err := client.CheckSecretEngineInstalled(expectedToken, "secret/", "kv")
+
+	// Assert
+	require.NoError(t, err)
+	assert.True(t, installed)
```
Comment (jim-wang-intel) [inline internal/pkg/vault/managment_test.go:240]: I've added more tests for check engine installed on consul as well, please modify this test based on that.
https://github.com/edgexfoundry/edgex-go/pull/3179/files#diff-87dbdd4059b58a71c40514e7ed80b05bf39658cdc7548d498004faad697397aa

File: internal/pkg/vault/managment_test.go [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+	defer ts.Close()
+
+	client := createClient(t, ts.URL, mockLogger)
+
+	// Act
+	installed, err := client.CheckSecretEngineInstalled(expectedToken, "secret/", "kv")
+
+	// Assert
+	require.NoError(t, err)
+	assert.False(t, installed)
```
Comment (jim-wang-intel) [inline internal/pkg/vault/managment_test.go:322]: I've added more tests for check engine NOT installed on consul as well, please modify this test based on that.
https://github.com/edgexfoundry/edgex-go/pull/3179/files#diff-87dbdd4059b58a71c40514e7ed80b05bf39658cdc7548d498004faad697397aa

File: internal/pkg/vault/models.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	Progress int  `json:"progress"`
+}
+
+type SecretsEngineOptions struct {
+	Version string `json:"version"`
+}
+
+// SecretsEngineConfig is config for /v1/sys/mounts
+type SecretsEngineConfig struct {
+	DefaultLeaseTTLDuration string
```
Comment (jim-wang-intel) [inline internal/pkg/vault/models.go:104]: add the json notation:
```suggestion
	DefaultLeaseTTLDuration string `json:"default_lease_ttl"`
```

File: internal/pkg/vault/request.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	// URL path
+	Path string
+	// If non-nil, passed to JSON serializer and included in request
+	JSONObject interface{}
+	// Included in HTTP request if JSONObject is nil
+	BodyReader io.Reader
+	// Description of the operation being performed included in log messages
+	OperationDescription string
+	// Expected status code to be returned from HTTP request
+	ExpectedStatusCode int
```
Comment (jim-wang-intel) [inline internal/pkg/vault/request.go:45]: I need to understand the background of this assumption more: Not sure this assumption is ok: in some cases, there may be more than one expected status codes from the client. One case is like `unseal` the Vault, 200 might not be the only expected code from the client point of view; maybe all status code in the range of 200-299 would be ok?

File: internal/pkg/vault/request.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	req.Header.Set("Content-Type", clients.ContentTypeJSON)
+	resp, err := c.HttpCaller.Do(req)
+
+	if err != nil {
+		c.lc.Error(fmt.Sprintf("unable to make request to %s failed: %s", params.OperationDescription, err.Error()))
+		return 0, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode != params.ExpectedStatusCode {
```
Comment (jim-wang-intel) [inline internal/pkg/vault/request.go:80]: should be in range?

File: pkg/requestor.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+func NewRequester(logger logger.LoggingClient) HTTPSRequester {
+	return &fluentRequester{logger}
+}
+
+func (r *fluentRequester) Insecure() Caller {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	return &http.Client{Timeout: 10 * time.Second, Transport: tr}
```
Comment (jim-wang-intel) [inline pkg/requestor.go:50]: prefer to define a constant for this `Timeout`:
```suggestion
	return &http.Client{Timeout: httpClientTimeoutDuration, Transport: tr}
```
where httpClientTimeoutDuration = 10 * time.Second

File: pkg/requestor.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs:            caCertPool,
+			InsecureSkipVerify: false,
+			ServerName:         serverName,
+		},
+		TLSHandshakeTimeout: 10 * time.Second,
```
Comment (jim-wang-intel) [inline pkg/requestor.go:71]: use constant

File: pkg/requestor.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	return &http.Client{Timeout: 10 * time.Second, Transport: tr}
```
Comment (jim-wang-intel) [inline pkg/requestor.go:73]: use constant

File: pkg/requestor.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+func NewMockRequester() *mockRequester {
+	return &mockRequester{}
+}
+
+func (r *mockRequester) Insecure() Caller {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	return &http.Client{Timeout: 10 * time.Second, Transport: tr}
```
Comment (jim-wang-intel) [inline pkg/requestor.go:87]: use constant

File: pkg/types/config.go [hunk 12]
```diff
@@ -38,25 +39,18 @@ type SecretConfig struct {
...
 
-	rawUrl := fmt.Sprintf("%s://%s:%v%s", c.Protocol, c.Host, c.Port, path)
-
-	u, err := url.Parse(rawUrl)
-	if err != nil {
-		return "", err
-	}
-
-	return u.String(), nil
+	return fmt.Sprintf("%s://%s:%v%s", c.Protocol, c.Host, c.Port, path)
```
Comment (jim-wang-intel) [inline pkg/types/config.go:48]: where the validation of url logic goes?

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]
Looks pretty good. Some suggestions to change.  Also, I've created secretengine.enabler like factory method to be hooked with `secretstore-setup`, should that part be in here or you prefer to leave it to the edgex-go level?

File: pkg/types/config.go (hunk 12 above)
Comment (lenny-intel) [reply pkg/types/config.go:48]: Not needed, we are building it be hand. The requestor displays the URL if there is an error.

File: internal/pkg/vault/request.go (hunk 6 above)
Comment (lenny-intel) [reply internal/pkg/vault/request.go:45]: This is the existing code from SecretStoreClient that is working. It is making a specific request with a specific known status code expected.

**Comment by lenny-intel** [comment]
> I've created secretengine.enabler like factory method to be hooked with `secretstore-setup`, should that part be in here or you prefer to leave it to the edgex-go level?

@jim-wang-intel , I don't understand the need for it, but it doesn't belong in this abstraction. So if really needed, then stays in edgex-go.

File: internal/pkg/vault/management.go (hunk 1 above)
Comment (lenny-intel) [reply internal/pkg/vault/management.go:142]: Humm, I had that... no sure where it went.

File: internal/pkg/vault/management.go (hunk 2 above)
Comment (lenny-intel) [reply internal/pkg/vault/management.go:183]: @jim-wang-intel , where are you getting `engine` from? This API is specific to Consul, but.. we could combine the two into `EnableSecretEngine` which is where I think you are going... ;-)

Comment (lenny-intel) [reply internal/pkg/vault/management.go:183]: Naw, that gets a little to messy... :-(

File: internal/pkg/vault/managment_test.go (hunk 3 above)
Comment (bnevis-i) [reply internal/pkg/vault/managment_test.go:240]: `managmetn_test` --> `management_test`

File: internal/pkg/vault/request.go (hunk 7 above)
Comment (lenny-intel) [reply internal/pkg/vault/request.go:80]: No this is the way this Requester works.

File: internal/pkg/vault/management.go (hunk 2 above)
Comment (lenny-intel) [reply internal/pkg/vault/management.go:183]: Yup, good catch.

File: pkg/types/config.go (hunk 12 above)
Comment (jim-wang-intel) [reply pkg/types/config.go:48]: I disagree, the Protocol string can be input from user and could be anything: why not validate?

**Comment by jim-wang-intel** [comment]
> > I've created secretengine.enabler like factory method to be hooked with `secretstore-setup`, should that part be in here or you prefer to leave it to the edgex-go level?
> 
> @jim-wang-intel , I don't understand the need for it, but it doesn't belong in this abstraction. So if really needed, then stays in edgex-go.

Ok, I'll leave it at edgex-go level, thanks!

File: internal/pkg/vault/request.go (hunk 6 above)
Comment (jim-wang-intel) [reply internal/pkg/vault/request.go:45]: ok, now I see this better: only applies to a single request in a http url call.

File: pkg/types/config.go (hunk 12 above)
Comment (lenny-intel) [reply pkg/types/config.go:48]: Ok

Comment (lenny-intel) [reply pkg/types/config.go]: More robust validation added back.

**Review by jim-wang-intel — CHANGES_REQUESTED** [review]
Very close: a couple of suggestions and typo change.

File: internal/pkg/vault/secrets_test.go [hunk 13]
```diff
@@ -503,7 +501,7 @@ func TestHttpSecretStoreManager_GetValue(t *testing.T) {
 			keys:              []string{"one"},
 			expectedValues:    nil,
 			expectError:       true,
-			expectedErrorType: &url.Error{},
+			expectedErrorType: nil,
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:504]: The added robust error message changes it from being a url.Error{} to a customer error so can't verify it as a url.Error{}.

File: internal/pkg/vault/secrets_test.go [hunk 14]
```diff
@@ -699,7 +694,7 @@ func TestHttpSecretStoreManager_SetValue(t *testing.T) {
 			secrets:           map[string]string{"one": "uno"},
 			expectedValues:    nil,
 			expectError:       true,
-			expectedErrorType: &url.Error{},
+			expectedErrorType: nil,
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:697]: Same repsonse

File: pkg/types/config.go [hunk 15]
```diff
@@ -39,18 +40,39 @@ type SecretConfig struct {
...
-	return fmt.Sprintf("%s://%s:%v%s", c.Protocol, c.Host, c.Port, path)
+	if len(c.Protocol) == 0 {
+		return "", fmt.Errorf("unable to build URL: Protocol not set. Please check configuration settings")
+	}
+
+	if len(c.Host) == 0 {
+		return "", fmt.Errorf("unable to build URL: Host not set. Please check configuration settings")
+	}
+
+	if c.Port == 0 {
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply pkg/types/config.go:57]: This is for un-initialized values. A negative value will be handled by the url.Parse().

File: secrets/factory.go [hunk 16]
```diff
@@ -16,102 +16,43 @@ package secrets
...
-	factory.mapMutex.Lock()
-	// if there is context already associated with the given token,
-	// then we cancel it first
-	if cancel, exists := factory.tokenCancelFunc[tokenStr]; exists {
-		cancel()
-	}
-	factory.mapMutex.Unlock()
+// NewManagementClient creates a new instance of a SecretClient based on the passed in configuration.
+// The NewSecretStoreClient provide management functionality to manage the secret store.
+func NewSecretStoreClient(config types.SecretConfig, lc logger.LoggingClient, requester pkg.Caller) (SecretStoreClient, error) {
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply secrets/factory.go:50]: Let over for first name I was using... ;-)

File: internal/pkg/vault/secrets_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/pkg/vault/secrets_test.go:504]: so....if that is the case, we should know the customer error type, don't we?

File: pkg/types/config.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply pkg/types/config.go:57]: ok

File: internal/pkg/vault/secrets_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:504]: No, it isn't a custom type., just a regular error with custom message.

[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/pkg/vault/secrets_test.go:504]: then just use `errors.New("")` as the expected type

[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:504]: That doesn't add anything. We already know it is an `error` and have verified that with `require.Error()`

[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/pkg/vault/secrets_test.go:504]: I know, but `nil` errortype doesn't make sense to me in this case

[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:504]: It simply indicates that there isn't a specific error type expected. i.e. any. The code only check for the specific error type if this is not nil. We use this approach is our tests all the time.

[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/pkg/vault/secrets_test.go:504]: I don't think that approach is right. if you have non-nil error, and expecting some error, you should get non-nil error type as expected error type.  The nil error type just doesn't fit here.  i am ok to put special logic with test name to skip the error type check if you think the error type is redundant. but sorry, `nil` error type just don't make sense to me....

[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/pkg/vault/secrets_test.go:504]: Ok, I can settle for `errors.New("") as the expected type`. ;-)

**Review by jim-wang-intel — APPROVED** [review]
LGTM

**Review by bnevis-i — APPROVED** [review]

**Comment by lenny-intel** [comment]
recheck

**Pull request merged**